src/charm.py

```diff
@@ -8,18 +8,223 @@
+                "version": {
+                    "override": "replace",
+                    "exec": {"command": "hydra version"},
+                },
+                "ready": {
+                    "override": "replace",
+                    "http": {"url": "http://localhost:4445/health/ready"},
+                },
+            },
+        }
+        return Layer(layer_config)
+
+    @property
+    def _config(self) -> str:
+        """Returns Hydra configuration."""
+        db_info = self._get_database_relation_info()
+
+        config = {
+            "dsn": f"postgres://{db_info['username']}:{db_info['password']}@{db_info['endpoints']}/postgres",
+            "log": {"level": "trace"},
+            "secrets": {
+                "cookie": ["my-cookie-secret"],
+                "system": ["my-system-secret"],
+            },
+            "serve": {
+                "admin": {
+                    "host": "localhost",
+                    "port": 4445,
+                },
+                "public": {
+                    "host": "localhost",
+                    "port": 4444,
+                },
+            },
+            "urls": {
+                "consent": "http://localhost:3000/consent",
+                "login": "http://localhost:3000/login",
+                "self": {
+                    "issuer": "http://localhost:4444/",
+                    "public": "http://localhost:4444/",
+                },
+            },
+        }
+
+        return yaml.dump(config)
+
+    def _check_leader(self) -> None:
+        if not self.unit.is_leader():
+            # We can't do anything useful when not the leader, so do nothing.
+            raise CheckFailedError("Waiting for leadership", WaitingStatus)
+
+    def _get_database_relation_info(self) -> dict:
+        relation_id = self.database.relations[0].id
+        relation_data = self.database.fetch_relation_data()[relation_id]
+
+        return {
+            "username": relation_data["username"],
+            "password": relation_data["password"],
+            "endpoints": relation_data["endpoints"],
+        }
+
+    def _update_layer(self) -> None:
+        """Updates the Pebble configuration layer and config if changed."""
+        # Get current layer
+        current_layer = self._container.get_plan()
+        # Create a new config layer
+        new_layer = self._hydra_layer
+
+        if current_layer.services != new_layer.services:
+            self.unit.status = MaintenanceStatus("Applying new pebble layer")
+            self._container.add_layer(self._container_name, new_layer, combine=True)
+            logger.info("Pebble plan updated with new configuration")
+            # Push the config on first layer update to avoid PathError
+            self._container.push(self._hydra_config_path, self._config, make_dirs=True)
+            logger.info("Pushed hydra config")
+
+        try:
+            current_config = self._container.pull(self._hydra_config_path).read()
```

Comment:
I thought we are pushing the config file when the layer is changed (ie first layer) instead of the whole try except block?

Reply:
That's correct, but we need to keep comparing the config file later when the pebble layer is in place and doesn't change, hence the try except block